garrysmod/lua/vgui/ddialogbox.lua

```diff
@@ -0,0 +1,203 @@
+	self.primaryPanel:DockPadding( 4, 4, 4, 4 ) -- Inner
+	self.primaryPanel:Dock( TOP )
+	self.primaryPanel:SetBackgroundColor( Color( 255, 0, 0, 0 ) )
+	table.insert( self.children, self.primaryPanel )
+	
+	-- Main read-only text content of the dialog
+	if #self.text > 0 then
+		-- Create the primary label, set text, and resize to fit content
+		self.label = vgui.Create( "DLabel", self.primaryPanel )
+		self.label:Dock( TOP )
+		self.label:DockMargin( 0, 0, 0, 4 ) -- Outer
+		self.label:SetText( self.text )
+		self.label:SizeToContents()
+		self.label:SetWrap( true )
+		self.label:SetAutoStretchVertical( true )
+	end
+	
+	self.secondaryPanel = vgui.Create( "DPanel", self )
+	self.secondaryPanel:DockMargin( 0, 0, 0, 0 ) -- Outer
+	self.secondaryPanel:DockPadding( 0, 0, 0, 0 ) -- Inner
+	self.secondaryPanel:Dock( TOP )
+	self.secondaryPanel:SetTall( 30 )
+	self.secondaryPanel:SetBackgroundColor( Color( 0, 255, 0, 0 ) )
+	table.insert( self.children, self.secondaryPanel )
+	
+	if self:GetRejectButtonEnabled() then
+		self.rejectButton = vgui.Create( "DButton", self.secondaryPanel )
+		self.rejectButton:SetText( self:GetRejectButtonText() )
+		self.rejectButton:SizeToContents()
+		local width = self.rejectButton:GetWide()
+		width = math.max( width + 20, 66 )
+		self.rejectButton:SetWide( width )
+		self.rejectButton:Dock( RIGHT )
+		self.rejectButton:DockMargin( 0, 4, 4, 4 ) -- Outer
+		self.rejectButton.DoClick = function()
+			self.handled = true
+			self:OnReject()
+			self:Close()
+		end
+	end
+	
+	if self:GetAcceptButtonEnabled() then
+		self.acceptButton = vgui.Create( "DButton", self.secondaryPanel )
+		self.acceptButton:SetText( self:GetAcceptButtonText() )
+		self.acceptButton:SizeToContents()
+		local width = self.acceptButton:GetWide()
+		width = math.max( width + 20, 66 )
+		self.acceptButton:SetWide( width )
+		self.acceptButton:Dock( RIGHT )
+		self.acceptButton:DockMargin( 0, 4, 4, 4 ) -- Outer
+		self.acceptButton.DoClick = function()
+			self.handled = true
+			self:OnAccept()
+			self:Close()
+		end
+	end
+	
+	-- Resize dialog frame to fit children
+	self:Invalidate()
+end
+
+function PANEL:GetText()
+	return self.text
+end
+
+function PANEL:SetText( value )
+	self.text = value
+	self:RebuildContents()
+end
+
+-- The accept button is the choice button that confirms the requested action
+function PANEL:GetAcceptButtonEnabled()
+	return self.isAcceptButtonEnabled
+end
+
+function PANEL:SetAcceptButtonEnabled( value )
+	self.isAcceptButtonEnabled = value
+	self:RebuildContents()
+end
+
+function PANEL:GetAcceptButtonText()
+	return self.acceptButtonText
+end
+
+function PANEL:SetAcceptButtonText( value )
+	self.acceptButtonText = value
+	self:RebuildContents()
+end
+
+-- The reject button is the choice button that denies the requested action
+function PANEL:GetRejectButtonEnabled()
+	return self.isRejectButtonEnabled
+end
+
+function PANEL:SetRejectButtonEnabled( value )
+	self.isRejectButtonEnabled = value
+	self:RebuildContents()
```

Comment:
I think this way of doing this unnecessarily recreates all the panels multiple times during panel creation, when you call multiple functions, like `SetRejectButtonEnabled`, `SetRejectButtonText`, etc. on the same panel.

I think creating your panels once in `PANEL:Init` and then toggling visibility/`SetText`ing in these setters would be a better idea, or doing something similar to how https://wiki.facepunch.com/gmod/PANEL:PerformLayout functions (i.e. using a 'dirty' flag and rebuilding only once per frame).